HEN_HOUSE/egs++/egs_rndm.cpp

```diff
@@ -404,6 +404,230 @@ void EGS_Ranmar::fillArray(int n, EGS_Float *array) {
+
+private:
+
+    void setState(int luxury, unsigned long int seedin);
+    unsigned long int incrState();
+    unsigned long int getSingle();
+
+    unsigned int i,j,n,skip,carry;
+    unsigned long int u[24];
+
+    unsigned int iseed1, iseed2;
+
+    static const unsigned long int mask_lo,mask_hi,twop24;
+    static const int skip_lux[5];
+    static const int next[24];
+
+    EGS_Ranlux *copy;
+
+};
+
+void EGS_Ranlux::saveState() {
+    if (copy) {
+        copy->set(*this);
+    }
+    else {
+        copy = new EGS_Ranlux(*this);
+    }
+    copy->copy = 0;
+}
+
+void EGS_Ranlux::resetState() {
+    if (copy) {
+        EGS_Ranlux *tmp = copy;
+        set(*copy);
+        copy = tmp;
+    }
+}
+
+EGS_RandomGenerator *EGS_Ranlux::getCopy() {
+    EGS_Ranlux *c = new EGS_Ranlux(*this);
+    c->np = 0;
+    c->copy = 0;
+    c->copyBaseState(*this);
+    return c;
+}
+
+void EGS_Ranlux::setState(EGS_RandomGenerator *r) {
+    copyBaseState(*r);
+    EGS_Ranlux *r1 = dynamic_cast<EGS_Ranlux *>(r);
+    if (!r1) {
+        egsFatal("EGS_Ranlux::setState: attempt to set my state by a non EGS_Ranlux RNG\n!");
+    }
+    i = r1->i;
+    j = r1->j;
+    n = r1->n;
+    skip = r1->skip;
+    carry = r1->carry;
+    for (int j=0; j<24; j++) {
+        u[j] = r1->u[j];
+    }
+}
+
+bool EGS_Ranlux::storePrivateState(ostream &data) {
+    data << i << " " << j << " " << n << " " << skip << " " << carry << endl;
+    for (int j=0; j<24; j++) {
+        data << u[j] << " ";
+    }
+    data << endl;
+    return data.good();
+}
+
+bool EGS_Ranlux::setPrivateState(istream &data) {
+    data >> i >> j >> n >> skip >> carry;
+    for (int j=0; j<24; j++) {
+        data >> u[j];
+    }
+    return data.good();
+}
+
+const unsigned long int EGS_Ranlux::mask_lo =  0x00ffffffUL;
+const unsigned long int EGS_Ranlux::mask_hi = ~0x00ffffffUL;
+const unsigned long int EGS_Ranlux::twop24  =  0x01000000UL;
+const int EGS_Ranlux::skip_lux[5] = {0,24,73,199,365}; /* p - 24 */
+const int EGS_Ranlux::next[24] = {23,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22};
+
+void EGS_Ranlux::describeRNG() const {
+    egsInformation("Random number generator:\n"
+                   "============================================\n");
+    egsInformation("  type                = ranlux\n");
+    egsInformation("  luxury level        = %d\n",iseed1);
+    egsInformation("  initial seed        = %d\n",iseed2);
+    egsInformation("  numbers used so far = %lld\n",count);
+}
+
+void EGS_Ranlux::setState(int luxury, unsigned long int seedin) {
+    unsigned long int k;
+    long int seed;
+    int j;
+
+    seed = seedin;
+    if( seed == 0 ) seed = 314159265;
+    if( luxury < 0 || luxury > 4 ) luxury = 1;
+
+    iseed1 = luxury;
+    iseed2 = seedin;
+
+    for (j=0; j<24; j++) {
+        k = seed / 53668;
+        seed = 40014 * (seed - k * 53668) - k * 12211;
+        if( seed < 0 ) seed += 2147483563;
+        u[j] = seed % twop24;
+    }
+
+    i = 23;
+    j = 9;
```

Comment:
There appears to be a variable shadowing bug here: The local `j` is set to 9, not `EGS_Ranlux::j` as likely intended (AddressSanitizer complains about an out of bounds read for `u[j]` in `incrState` later on with the current version). Changing the local `j` to `jj` seems to run OK.